File: question3/question3.js
// Let's say you have a list of N+1 integers between 1 and N. You know there's at least one
// duplicate, but there might be more. For example, if N=3, your list might be 3, 1, 1, 3 or it might
// be 1, 3, 2, 2. Print out a number that appears in the list more than once. (That is, in the first
// example, you can print '1' or '3' -- you don't have to print both.)
// Program Input: A list of numbers between 0 and N, all integers, comma separated.
// Program Output: A list of all numbers that are duplicated in the list.

function findDuplicates(numberArr) {
  let obj = {};
  let result = [];

  for (let i = 0; i < numberArr.length; i++) {
    if (obj[numberArr[i]]) {
      obj[numberArr[i]]++;
    } else {
      obj[numberArr[i]] = 1;
    }
  }
  for (let key in obj) {
    if (obj[key] >= 2) {
      result.push(Number(key));
    }
  }
  return result;
}

File: question3/question3.cs
using System;
using System.Collections.Generic;

namespace question3
{
    class Program
    {
        public static void Main(string[] args)
        {
            Console.Write("Enter a series of numbers separated by commas: ");
            string numbersInput = Console.ReadLine();
            string[] stringArr = numbersInput.Split(',');
            int[] numbersArr = Array.ConvertAll(stringArr, int.Parse);
            Console.WriteLine("Repeated numbers: ");
            Console.WriteLine(string.Join(", ", FindDuplicates(numbersArr).ToArray()));
        }
        public static List<int> FindDuplicates(int[] numberArr)
        {
            // create a cache dictionary
            Dictionary<int, int> cache = new Dictionary<int, int>();
            var result = new List<int>();
            // map the number as the key of the obj and increase the count of the value by 1
            for (int i = 0; i < numberArr.Length; i++)
            {
                if (cache.ContainsKey(numberArr[i]))
                {
                    cache[numberArr[i]]++;
                }
                else
                {
                    cache[numberArr[i]] = 1;
                }
            }
            // loop over the cache and add all the ones that are duplicated to result
            foreach (var key in cache.Keys)
            {
                if (cache[key] >= 2)
                {
                    result.Add(key);
                }
            }
            return result;
        }
    }
}


